File: src/examples/ex_pick.cpp
/* ex_pick.cpp
 * Scripted pick: Kinova GEN3 + Robotiq 2F-85 picks an orange cube from the floor and
 * lifts it.
 *
 * Joint impedance control law applied every step:
 *   tau_i = g_kdl_i + Kp*(q_des_i - q_i) - Kd*dq_i
 * where g_kdl is computed via KDL::ChainDynParam::JntToGravity (includes gripper inertia).
 *
 * Phases: HOME -> PREGRASP -> GRASP -> CLOSE -> LIFT -> HOLD
 *
 * Requires MuJoCo Menagerie in cache.
 *
 * Usage:
 *   ex_pick [--headless]
 *
 * With --headless runs the full pick sequence and prints final cube height. */

#include "mj_kdl_wrapper/mj_kdl_wrapper.hpp"
#include "common.hpp"
#include "example_paths.hpp"

#include <kdl/chaindynparam.hpp>
#include <kdl/chainfksolverpos_recursive.hpp>
#include <kdl/chainiksolvervel_wdls.hpp>

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

using mj_kdl_examples::kGripperClosed;

// scene constants
using mj_kdl_examples::kHomePose;
static constexpr double kCubeX       = 0.4;
static constexpr double kCubeY       = 0.0;
static constexpr double kCubeHS      = 0.02; // half-size of 4 cm cube
static constexpr double kCubeZ       = kCubeHS;

// impedance gains (per joint, matching ex_impedance tuning)
static constexpr double kKp[7] = { 100, 200, 100, 200, 100, 200, 100 }; // Nm/rad
static constexpr double kKd[7] = { 10, 20, 10, 20, 10, 20, 10 };        // Nm*s/rad

/*
 * Compute joint impedance torques into robot.jnt_trq_cmd:
 *   tau_i = g_kdl[i] + Kp[i]*(q_des_i - q_msr_i) - Kd[i]*dq_msr_i
 * dyn must have been built from the chain returned by init_robot_from_mjcf with tool_body set.
 */
static void impedance_ctrl(
  mj_kdl::Robot       &robot,
  const KDL::JntArray &q_des,
  unsigned             n,
  KDL::ChainDynParam  &dyn
)
{
    KDL::JntArray q(n), g(n);
    for (unsigned i = 0; i < n; ++i) q(i) = robot.jnt_pos_msr[i];
    dyn.JntToGravity(q, g);
    for (unsigned i = 0; i < n; ++i) {
        robot.jnt_trq_cmd[i] =
          g(i) + kKp[i] * (q_des(i) - robot.jnt_pos_msr[i]) - kKd[i] * robot.jnt_vel_msr[i];
    }
}

int main(int argc, char *argv[])
{
    const bool headless = mj_kdl_examples::parse_args(argc, argv).headless;

    // scene setup
    const std::string arm_mjcf = mj_kdl_examples::menagerie_model("kinova_gen3/gen3.xml");
    const std::string grp_mjcf = mj_kdl_examples::asset("robotiq_2f85/2f85.xml");

    mj_kdl::AttachmentSpec gs;
    gs.mjcf_path = grp_mjcf;
    gs.attach_to = { mj_kdl::AttachKind::Site, "pinch_site" };
    gs.prefix    = "g_";

    mj_kdl::RobotSpec rs;
    rs.path = arm_mjcf;
    rs.attachments.push_back(gs);

    mj_kdl::SceneObject cube{
        .name      = "cube",
        .mjcf_path = "",
        .shape     = mj_kdl::Shape::BOX,
        .size      = { kCubeHS, kCubeHS, kCubeHS },
        .pos       = { kCubeX, kCubeY, kCubeZ },
        .rgba      = { 1.0f, 0.5f, 0.0f, 1.0f },
        .mass      = 0.1,
        .condim    = mj_kdl::Condim::Torsional,
        .friction  = { 0.8, 0.02, 0.001 },
    };

    mj_kdl::SceneSpec sc;
    sc.timestep   = 0.002;
    sc.add_floor  = true;
    sc.add_skybox = true;
    sc.robots.push_back(rs);
    sc.objects.push_back(cube);

    mj_kdl::Env env;
    if (!mj_kdl::init_env(&env, &sc)) {
        std::cerr << "init_env() failed\n";
        return 1;
    }
    const mjModel *model = env.model;

    mj_kdl::ToolFrameSpec tool;
    tool.tool_body = "g_base_mount";
    tool.tcp_site  = "g_pinch";

    mj_kdl::Robot robot;
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link", "", &tool)) {
        std::cerr << "init_robot_from_mjcf() failed\n";
        return 1;
    }

    unsigned                   n       = robot.chain.getNrOfJoints();
    mj_kdl::SceneActuatorSlot *fingers = mj_kdl::bind_scene_actuator(&env.scene, "g_fingers_actuator");
    int                        cube_jnt = mj_name2id(model, mjOBJ_JOINT, "cube_joint");
    int                        key_id   = mj_name2id(model, mjOBJ_KEY, "home");
    if (!fingers) return 1;

    KDL::ChainDynParam dyn(robot.chain, KDL::Vector(0.0, 0.0, -9.81));

    // IK setup
    KDL::ChainFkSolverPos_recursive fk(robot.chain);
    KDL::ChainIkSolverVel_wdls      ik_vel(robot.chain, 1e-5, 150);
    ik_vel.setLambda(0.05);

    KDL::JntArray q_home(n), q_pregrasp(n), q_grasp(n), q_lift(n);
    for (unsigned i = 0; i < n; ++i) q_home(i) = kHomePose[i];
    const double        kGraspZ    = kCubeZ;
    const double        kPreGraspZ = kGraspZ + 0.20;
    const double        kLiftZ     = kGraspZ + 0.30;
    const KDL::Rotation kGraspRot  = robot.tip_T_tcp.M;

    struct WP
    {
        double               z;
        KDL::JntArray       *out;
        const KDL::JntArray *seed;
    };
    WP wps[] = {
        { kPreGraspZ, &q_pregrasp, &q_home },
        { kGraspZ, &q_grasp, &q_pregrasp },
        { kLiftZ, &q_lift, &q_grasp },
    };
    for (auto &wp : wps) {
        KDL::Frame target(kGraspRot, KDL::Vector(kCubeX, kCubeY, wp.z));
        if (!mj_kdl_examples::solve_near_seed(ik_vel, fk, robot, *wp.seed, target, *wp.out)) {
            std::cerr << "IK failed for z=" << wp.z << "\n";
            return 1;
        }
    }

    const double kHoldDuration = 1.0;

    // clang-format off
    const std::vector<mj_kdl_examples::Phase> phases = {
        { "HOME",     &q_home,     1.0,           2.5,            0.08, 0.0            },
        { "PREGRASP", &q_pregrasp, 5.0,           7.0,            0.08, 0.0            },
        { "GRASP",    &q_grasp,    5.0,           8.0,            0.03, 0.0            },
        { "CLOSE",    &q_grasp,    1.5,           2.5,           -1.0,  kGripperClosed },
        { "LIFT",     &q_lift,     3.0,           5.0,            0.08, kGripperClosed },
        { "HOLD",     &q_lift,     kHoldDuration, kHoldDuration, -1.0,  kGripperClosed },
    };
    // clang-format on

    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::TORQUE)) return 1;

    mj_kdl::ResetOptions reset_opts;
    reset_opts.use_keyframe = key_id >= 0;
    reset_opts.keyframe     = key_id >= 0 ? key_id : 0;

    bool restart = false;
    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        if (key_id < 0) mj_kdl::set_joint_pos(&robot, q_home);
        double *cube_q = ctx->data->qpos + model->jnt_qposadr[cube_jnt];
        cube_q[0]      = kCubeX;
        cube_q[1]      = kCubeY;
        cube_q[2]      = kCubeZ;
        cube_q[3]      = 1.0;
        cube_q[4] = cube_q[5] = cube_q[6] = 0.0;

        ctx->data->ctrl[fingers->ctrl_id] = 0.0;
        restart                           = true;
    };

    mj_kdl::reset(&env, &reset_opts);

    if (!headless && !mj_kdl::open_viewer(&env)) {
        std::cerr << "open_viewer() failed\n";
        return 1;
    }

    mj_kdl_examples::run_phases(
      env,
      robot,
      fingers,
      phases,
      restart,
      [&](const KDL::JntArray &q_des) { impedance_ctrl(robot, q_des, n, dyn); }
    );

    if (headless) {
        int    qadr   = model->jnt_qposadr[cube_jnt];
        double cube_z = env.data->qpos[qadr + 2];
        std::cout << "cube Z after pick: " << std::fixed << std::setprecision(3) << cube_z
                  << " m\n";
    }

    mj_kdl::cleanup(&env);
    return 0;
}

File: python/examples/ex_pick.py
#!/usr/bin/env python3
"""Scripted floor pick example ported from src/examples/ex_pick.cpp.

Joint-impedance control (tau = Kp*e - Kd*qdot + g_kdl) drives scripted
waypoints to grasp and lift a cube. An Env on_reset hook re-homes the arm,
re-poses the cube, and opens the gripper, so the simulate-UI reset button
replays the whole pick.
"""

from __future__ import annotations

import argparse
import math
from dataclasses import dataclass

import PyKDL as kdl
import mj_kdl_wrapper as mjk

HOME = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
CUBE_X = 0.40
CUBE_Y = 0.00
CUBE_HALF_SIZE = 0.02
CUBE_Z = CUBE_HALF_SIZE
KP = [100.0, 200.0, 100.0, 200.0, 100.0, 200.0, 100.0]
KD = [10.0, 20.0, 10.0, 20.0, 10.0, 20.0, 10.0]


class ResetRequested(Exception):
    """Raised when the simulate UI reset is detected, to restart the sequence."""


@dataclass(frozen=True)
class Phase:
    name: str
    target: list[float]
    duration: float
    timeout: float
    settle_tol: float
    gripper: float


def gripper_attachment(gripper_path: str) -> mjk.AttachmentSpec:
    attach = mjk.AttachmentSpec()
    attach.mjcf_path = gripper_path
    attach.attach_to = mjk.AttachTarget(mjk.AttachKind.Site, "pinch_site")
    attach.prefix = "g_"
    return attach


def cube_object() -> mjk.SceneObject:
    cube = mjk.SceneObject()
    cube.name = "cube"
    cube.shape = mjk.Shape.BOX
    cube.size = [CUBE_HALF_SIZE, CUBE_HALF_SIZE, CUBE_HALF_SIZE]
    cube.pos = [CUBE_X, CUBE_Y, CUBE_Z]
    cube.rgba = [1.0, 0.5, 0.0, 1.0]
    cube.mass = 0.1
    cube.condim = mjk.Condim.Torsional
    cube.friction = [0.8, 0.02, 0.001]
    return cube


def build_env() -> tuple[mjk.Env, mjk.Robot]:
    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    spec.objects = [cube_object()]

    robot_spec = mjk.RobotSpec()
    robot_spec.path = mjk.menagerie.model_path("kinova_gen3", env_var="MJ_KDL_MODEL")
    robot_spec.attachments = [
        gripper_attachment(
            mjk.menagerie.asset_path("robotiq_2f85/2f85.xml", env_var="MJ_KDL_GRIPPER")
        )
    ]
    spec.robots = [robot_spec]

    env = mjk.Env.build(spec)
    tool = mjk.ToolFrameSpec()
    tool.tool_body = "g_base"
    tool.tcp_site = "g_pinch"
    robot = env.create_robot("base_link", "bracelet_link", tool=tool)
    return env, robot


def jnt(values: list[float]) -> kdl.JntArray:
    q = kdl.JntArray(len(values))
    for i, value in enumerate(values):
        q[i] = value
    return q


def as_list(q: kdl.JntArray) -> list[float]:
    return [q[i] for i in range(q.rows())]


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def clamp_joint(value: float, limit: tuple[float, float]) -> float:
    low, high = limit
    if math.isfinite(low) and math.isfinite(high) and high > low:
        return clamp(value, low, high)
    return value


def solve_position_ik(
    chain: kdl.Chain,
    seed_values: list[float],
    target: kdl.Frame,
    joint_limits: list[tuple[float, float]],
) -> list[float]:
    fk = kdl.ChainFkSolverPos_recursive(chain)
    ik = kdl.ChainIkSolverVel_wdls(chain)
    ik.setLambda(0.05)
    q = jnt(seed_values)
    dq = kdl.JntArray(q.rows())
    for _ in range(700):
        current = kdl.Frame()
        fk.JntToCart(q, current)
        dx = kdl.diff(current, target)
        pos_err = dx.vel.Norm()
        rot_err = dx.rot.Norm()
        if pos_err <= 0.003 and rot_err <= 0.03:
            return as_list(q)
        if pos_err > 0.05:
            dx.vel = dx.vel * (0.05 / pos_err)
        if rot_err > 0.20:
            dx.rot = dx.rot * (0.20 / rot_err)
        if ik.CartToJnt(q, dx, dq) < 0:
            raise RuntimeError("PyKDL IK velocity step failed")
        for i in range(q.rows()):
            q[i] = clamp_joint(q[i] + dq[i], joint_limits[i])
    raise RuntimeError("PyKDL IK did not converge")


def build_waypoints(env: mjk.Env, robot: mjk.Robot) -> dict[str, list[float]]:
    chain = robot.kdl_chain()
    target_rot = (env.body_frame("bracelet_link").Inverse() * env.site_frame("g_pinch")).M
    seed = HOME[:]

    def solve(x: float, y: float, z: float) -> list[float]:
        nonlocal seed
        seed = solve_position_ik(
            chain, seed, kdl.Frame(target_rot, kdl.Vector(x, y, z)), robot.joint_limits
        )
        return seed[:]

    return {
        "home": HOME[:],
        "pregrasp": solve(CUBE_X, CUBE_Y, CUBE_Z + 0.20),
        "grasp": solve(CUBE_X, CUBE_Y, CUBE_Z),
        "lift": solve(CUBE_X, CUBE_Y, CUBE_Z + 0.30),
    }


def apply_pd_gravity(robot: mjk.Robot, target: list[float]) -> None:
    robot.update()
    gravity = robot.gravity_torques(-9.81)
    robot.jnt_trq_cmd = [
        KP[i] * (target[i] - robot.jnt_pos_msr[i]) - KD[i] * robot.jnt_vel_msr[i] + gravity[i]
        for i in range(robot.n_joints)
    ]


def max_abs_joint_err(robot: mjk.Robot, target: list[float]) -> float:
    return max(abs(target[i] - robot.jnt_pos_msr[i]) for i in range(robot.n_joints))


def lerp(start: list[float], target: list[float], alpha: float) -> list[float]:
    return [a + alpha * (b - a) for a, b in zip(start, target)]


def step_once(
    env: mjk.Env, robot: mjk.Robot, viewer: mjk.SimulateViewer | None, state: dict
) -> bool:
    ok = viewer.step() if viewer is not None else robot.step()
    if not ok:
        return False
    if viewer is not None and env.time() < state["prev"] - 1e-6:  # UI reset pressed
        env.reset()
        state["prev"] = env.time()
        raise ResetRequested()
    state["prev"] = env.time()
    return True


def run_phase(
    env: mjk.Env,
    robot: mjk.Robot,
    phase: Phase,
    viewer: mjk.SimulateViewer | None,
    state: dict,
) -> bool:
    print(f"State: {phase.name}")
    robot.update()
    start = robot.jnt_pos_msr[:]
    t0 = env.time()
    while True:
        elapsed = env.time() - t0
        alpha = clamp(elapsed / phase.duration, 0.0, 1.0) if phase.duration > 0.0 else 1.0
        apply_pd_gravity(robot, lerp(start, phase.target, alpha))
        if env.has_actuator("g_fingers_actuator"):
            env.set_actuator_ctrl("g_fingers_actuator", phase.gripper)

        done_time = elapsed >= phase.duration
        done_pose = phase.settle_tol < 0.0 or max_abs_joint_err(robot, phase.target) <= phase.settle_tol
        done_timeout = phase.timeout > 0.0 and elapsed >= phase.timeout
        if (done_time and done_pose) or done_timeout:
            return True
        if viewer is not None and not viewer.is_running():
            return False
        if not step_once(env, robot, viewer, state):
            return False


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    parser.add_argument("--headless", action="store_true", help=argparse.SUPPRESS)
    args = parser.parse_args()

    env, robot = build_env()
    try:
        robot.ctrl_mode = mjk.CtrlMode.TORQUE

        def on_reset(ctx):
            robot.set_joint_pos(HOME, call_forward=False)
            env.set_body_pose("cube", [CUBE_X, CUBE_Y, CUBE_Z])
            if env.has_actuator("g_fingers_actuator"):
                env.set_actuator_ctrl("g_fingers_actuator", 0.0)

        env.on_reset = on_reset
        env.reset()

        waypoints = build_waypoints(env, robot)
        phases = [
            Phase("HOME", waypoints["home"], 1.0, 2.5, 0.08, 0.0),
            Phase("PREGRASP", waypoints["pregrasp"], 5.0, 7.0, 0.08, 0.0),
            Phase("GRASP", waypoints["grasp"], 5.0, 8.0, 0.03, 0.0),
            Phase("CLOSE", waypoints["grasp"], 1.5, 2.5, -1.0, 255.0),
            Phase("LIFT", waypoints["lift"], 3.0, 5.0, 0.08, 255.0),
            Phase("HOLD", waypoints["lift"], 1.0 if not args.gui else 10.0, 0.0, -1.0, 255.0),
        ]
        state = {"prev": env.time()}
        if args.gui:
            viewer = mjk.SimulateViewer.open(robot, "ex_pick.py")
            try:
                while viewer.is_running():
                    try:
                        for phase in phases:
                            if not run_phase(env, robot, phase, viewer, state):
                                raise StopIteration
                        break
                    except ResetRequested:
                        continue
                    except StopIteration:
                        break
            finally:
                viewer.close()
        else:
            for phase in phases:
                if not run_phase(env, robot, phase, None, state):
                    break
        cube = env.body_frame("cube")
        print(f"final cube height: {cube.p.z():.3f} m")
    finally:
        env.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
